### PR TITLE
Deliver pd emails immediately rather than queueing first

### PR DESCRIPTION
**Pull request by megcrenshaw**
Rather than sending emails from the cron job, this moves the sending of emails into the method that is called by the cron job –– this allows us to avoid "fake-sending" emails to test the method's behavior and moves all the actions into one place.

I looked at other cron jobs, and it looks like only this one has the pattern of sending mailers from the cron job itself.

## Testing story

Updated existing unit tests

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by TurnerRiley — APPROVED** [review]
This looks great!

**Comment by davidsbailey** [comment]
apologies for the slow response. there are two things on my mind:
1. how is the test coverage for this PR -- how much of the cronjob code, if any, is untested? is it possible to test locally, and if not what is the plan around when to deploy this change as well as when and what to look out for in terms of potential issues in prod?
2. should we / could we do something similar for `process_pd_workshop_ends` ? this doesn't need to happen in this PR and you aren't required to come up with a firm answer, but I'm curious if you had any insight based on your exposure to these codepaths.

**Comment by megcrenshaw** [comment]
> apologies for the slow response. there are two things on my mind:
> 
> 1. how is the test coverage for this PR -- how much of the cronjob code, if any, is untested? is it possible to test locally, and if not what is the plan around when to deploy this change as well as when and what to look out for in terms of potential issues in prod?
> 2. should we / could we do something similar for `process_pd_workshop_ends` ? this doesn't need to happen in this PR and you aren't required to come up with a firm answer, but I'm curious if you had any insight based on your exposure to these codepaths.

Great questions.

(1) You can see that there is still a fair bit of code in the cron job that is not tested –– and looking at it now, I'm realizing I messed up with this refactor, as these queued emails are never sent:
https://github.com/code-dot-org/code-dot-org/blob/77a750aa8ace24c1a2d688dd21ff2948b8f9c5e6/dashboard/bin/scheduled_pd_application_emails#L10-L16

I didn't even realize these get sent every day! That's a lot of reminders –– @dmcavoy, it's intentional that the reminder to teachers that their admin hasn't approved their application gets sent each day?

Regardless, there other parts of the cron job that aren't tested (e.g. that we're actually calling the method that does this) https://github.com/code-dot-org/code-dot-org/blob/77a750aa8ace24c1a2d688dd21ff2948b8f9c5e6/dashboard/bin/scheduled_pd_application_emails#L27.

I'm going to follow-up with a smaller PR that (a) puts the *sending* (not the queuing) of teacher reminder emails in a separate file with some simple tests, and (b) avoids the method call within the cron job. Once that's approved, I'll go in and send unique emails in the queue to get those out.

(2) Good call re: `process_pd_workshop_ends`. I actually think that script looks great in the sense that the cron job only calls a method from the model:
https://github.com/code-dot-org/code-dot-org/blob/77a750aa8ace24c1a2d688dd21ff2948b8f9c5e6/dashboard/bin/process_pd_workshop_ends

Unfortunately, the `process_ends` method itself is not actually tested. Ticket for that here: https://codedotorg.atlassian.net/browse/ACQ-503

**Comment by dmcavoy** [comment]
@megcrenshaw It is not intentional for us to be reminding teachers that they need admin approval every day. That sounds really annoying for our teachers. In general our reminder emails follow a weekly cadence.

**Comment by megcrenshaw** [comment]
> @megcrenshaw It is not intentional for us to be reminding teachers that they need admin approval every day. That sounds really annoying for our teachers. In general our reminder emails follow a weekly cadence.

I just went through the logic more carefully –– we only send these once, and only if it's been 6 days or more since the first was sent. So I think we're good here! But still, modifying the script to make this more clear.

**Comment by davidsbailey** [comment]
> (2) Good call re: `process_pd_workshop_ends`. I actually think that script looks great in the sense that the cron job only calls a method from the model: https://github.com/code-dot-org/code-dot-org/blob/77a750aa8ace24c1a2d688dd21ff2948b8f9c5e6/dashboard/bin/process_pd_workshop_ends
> 
> Unfortunately, the `process_ends` method itself is not actually tested. Ticket for that here: https://codedotorg.atlassian.net/browse/ACQ-503

Thanks for tracking this! My other question was, could we be sending some of these "workshop ends" emails right away, rather than waiting for a cronjob to send them? caveat: I'm asking without having fully dug in... the answer can be "no" 😅

**Comment by megcrenshaw** [comment]
Oh right, meant to answer that part! According to this part of the code
https://github.com/code-dot-org/code-dot-org/blob/98bdd54a8d157bd853be1c48460167baa50d6f6d/dashboard/app/models/pd/workshop.rb#L402-L409

it's not possible. I haven't looked more into it, or seen if it'd be possible with updated versions –– I think that'd involve a conversation with infra?

**Comment by megcrenshaw** [comment]
The attachment is for the CSF certificate, so perhaps if we no longer need that, we could change them to send right away
https://github.com/code-dot-org/code-dot-org/blob/98bdd54a8d157bd853be1c48460167baa50d6f6d/dashboard/app/mailers/pd/workshop_mailer.rb#L247-L257

Another option is to run the cron job specifically for these emails, but send them right away for all other courses.

**Comment by davidsbailey** [comment]
thank you for this investigation. I am learning a lot about our email system!

this is a great list of options. I was thinking about this primarily from a point of eliminating complexity / low hanging fruit as we see it, so I'm a bit worried about letting this balloon into a more complicated solution unless we want to reduce the latency on these particular emails as a product goal. 

@dmcavoy do you think the work to send workshop exit surveys sooner / immediately is worth tracking from a product perspective? here is some past discussion on how important (or not) it is to send exit surveys promptly: https://codedotorg.slack.com/archives/CL0LNHHN2/p1656523213488319

**Pull request merged**
